### DISCUSSION TITLE
Mirth Connect Public Comment

### DISCUSSION THREAD

**Discussion by twest-mirthconnect** [nextgenhealthcare/connect#5347] [Mirth Connect Roadmap]
### Product Release Planning

Beginning in September, the Mirth Connect Product Team will begin publishing their features for an upcoming major and minor release (no patch releases will be published ahead of time). These publications will be open to feedback, discussion, and follow-up from anyone using Mirth Connect. 

There will be references to polls, mock-ups, invitations to development efforts, possible design considerations, and more.

We hope this will improve our product release plans and the needs of our amazing community members and the businesses they support.

Thanks again for all the great feedback.

Sincerely,
**Travis West, Mirth Interop Product Management**
_Mirth Connect | Mirth Cloud Connect | Mirth Training and Education | Mirth Professional Services | Mirth Technical Support_

twest@nextgen.com

**Comment by twest-mirthconnect** [comment]
### Public Comment and Feedback

The following items are expected to be released with 4.2 (in mid-November ) or 4.3 (March). 

Please note that each release is subject to change and that feedback for items may impact the priority and potential development needed as we review and improve our roadmap, feature, bug, enhancement, or functionality. Not all work completed by the team will be represented below (Non-functional Requirements, Security Updates, Etc.).

_Note: The development team is working through CURES Certification in Q3 and Q4 of 2022._

**Comment by twest-mirthconnect** [comment]
- RS-8013: Prevent users from accidentally overwriting channel changes:
**Description**
In some cases, users can overwrite changes to a channel without knowing. 
For instance, if User A starts editing a channel, User B saves changes to the channel, then User A saves their changes, the channel will not have any of User B's changes. There should be a way for User A to know that he is going to overwrite changes that have happened since he started editing.
**Acceptance**
A user is notified when they may be overwriting changes to a channel that happened between them opening the channel for editing and saving their own changes.

**Comment by twest-mirthconnect** [comment]
- RS-8355: Simplify the process of selecting a listener port
**Description**
It can sometimes be difficult to find an open port when creating and deploying a new channel. We should make it easier for users to find an open port when configuring their channels.
**Acceptance Criteria**
Users can see a list of used ports within the channel configuration screen
Used ports means ports assigned to any listening connector which have a configurable port what ever their status is
Show channel status

**Reply by pacmano1** [reply]
And not in the Channels and Dashboard view as an extra column?

**Reply by tonygermano** [reply]
This is issue https://github.com/nextgenhealthcare/connect/issues/2115.

The issue @pacmano1 is referring to is https://github.com/nextgenhealthcare/connect/issues/5152.

@twest-mirthconnect I think a better approach than copying everything to a discussion would be to update the actual issue, and add the issue to the [4.2.0 milestone](https://github.com/nextgenhealthcare/connect/milestone/74) (which already exists and contains RS-6425 that you mention in your next comment.)

github provides tools that can cut down on your manual work and make things easier to find for everyone.

**Reply by jonbartels** [reply]
Wow, #2115 is ten years old. 

I've done related work under https://github.com/jonbartels/mirth-connect-dashboard-count . It's licensed under MPL 2.0 so feel free to use it under the terms of that license. 

From the README: 
> In the Channel list: Coming soon! There's an MC bug that prevents this from working but the plugin code is all ready to go.

So if NextGen intends to implement this listing feature as a Channel Column Plugin you'll need to fix https://github.com/nextgenhealthcare/connect/issues/5171 . Once 5171 is fixed the implementation of RS-8355 is relatively easy.

**Reply by twest-mirthconnect** [reply]
Thanks, Gents. I've shared the feedback with the team.

**Comment by twest-mirthconnect** [comment]
- RS-6425: JMS Connectors may fail to load classes in resources
**Description**
In some cases, JMS connectors will fail to find classes in resources that should be available to them.
We have connectors that will work with the JMS protocol but we don't have libraries for all of JMS so users usually have to import libraries to implement.  The reason we do not include them is probably size, maybe licensing.  Users create resources, add it to channel, etc.  Right now it does not work properly.  Libraries are getting removed after channel starts but they need to persist.
**Acceptance Criteria**
JMS Connectors should not fail when using library resources.

**Reply by jonbartels** [reply]
Does this address https://github.com/nextgenhealthcare/connect/issues/5399 ?

**Reply by JackieK5** [reply]
Yes, this addresses #5399.  We will be sure to include the links (when applicable) in future posts.

**Comment by twest-mirthconnect** [comment]
- RS-9218: Add User to Revert in Channel History - for New Channel Edit Dialog
**Description**
When adding dialogue to Channel Edit screen, the user is now saved in the metadata for a new message that includes the user in the message. The revert in the channel history now fails because it needs the user id for the revert. There may need to be more research though. 
**Acceptance Criteria**
The idea is to save the user id with the revert so that if a user is editing a channel and the revert is done, the user editing the channel will see the user that reverted the channel.

**Comment by twest-mirthconnect** [comment]
- RS-9220: Add new X-Context Header in Health Data Hub Sender Connector Type
**Description**
The addition of a new header in the HHD Sender connector type (also requested in Git Hub: .   They are working on external systems/third party application integrations with HDH using the Health Data Hub Sender connector type while configuring the Mirth Connect channel.  During this process, they have found that they need some information that is currently not available, the “X-context”.  They have requested a new configurable field be added and the data entered in that field be sent as the “X-context” header.
**Acceptance Criteria**
The X-context option is included as a fully configurable field of the HDH Sender
The X-context entry will be sent, as entered, in the headers

**Comment by twest-mirthconnect** [comment]
- RS-6376: Improve halt-ability of the File Reader/Writer
**Description**
If a network partition or something similar causes one of the file connection methods to hang forever, there is basically no recourse at that point. This happened with a customer, where a File Reader was stuck at the Linux OS layer listing files. Perhaps we could modify our file connectors so that *any time* a call to the file connection (whether it's local file, FTP, SFTP, etc.) is made, we first wrap that in a separate thread and call that instead. That way if it gets stuck and the channel has to be halted, we can actually properly respond to the halt and allow the channel to be shut down. The caveat is, that thread will still be out there lingering forever, until the server is restarted.
**Acceptance Criteria**
Allow the user to halt the channel that has hung

**Reply by jonbartels** [reply]
Discussion in Slack https://mirthconnect.slack.com/archives/C02SW0K4D/p1664987327724709

**Reply by dfauraz** [reply]
Also maybe review the try/catch of the file connectors. On errors like timeout there is not even an error in the logs, which makes it difficult to troubleshoot JCIFS issues for example. Of course that's a poller so the channel is going to try again, but without logs it's difficult to explain

**Comment by vinwi4** [comment]
Thank you for taking this initiative.  It will be make the roadmap clear with improved involvement from the community

**Comment by twest-mirthconnect** [comment]
### Update:

As we all know, change is constant so in this case, we are excited to improve the way we share both what we are doing and what we have done. 

**If something is done:**
If the development team has completed a ticket and committed the changes during the sprint, we now include it here (see our development commits Git Hub) for everyone to take a look, comment, find bugs, suggest improvements, etc.) We have already seen some very promising responses.

**If something is going to be done:**
For each release, I will post the new items (here) ahead of time so that everyone will have a chance to see what we are targeting. This will help us learn what your concerns, questions, and comments are before we develop them. 

I hope this helps everyone along the way when looking to Mirth Connect for product improvements.

Thanks,
Travis West
Product Manager, Mirth Connect
